examples/petsc/bpsdmplex.c

```diff
@@ -0,0 +1,747 @@
+  // Read CL options
+  ierr = PetscOptionsBegin(comm, NULL, "CEED BPs in PETSc", NULL); CHKERRQ(ierr);
+  bpChoice = CEED_BP1;
+  ierr = PetscOptionsEnum("-problem",
+                          "CEED benchmark problem to solve", NULL,
+                          bpTypes, (PetscEnum)bpChoice, (PetscEnum*)&bpChoice,
+                          NULL); CHKERRQ(ierr);
+  vscale = bpOptions[bpChoice].vscale;
+  test_mode = PETSC_FALSE;
+  ierr = PetscOptionsBool("-test",
+                          "Testing mode (do not print unless error is large)",
+                          NULL, test_mode, &test_mode, NULL); CHKERRQ(ierr);
+  benchmark_mode = PETSC_FALSE;
+  ierr = PetscOptionsBool("-benchmark",
+                          "Benchmarking mode (prints benchmark statistics)",
+                          NULL, benchmark_mode, &benchmark_mode, NULL);
+  CHKERRQ(ierr);
+  qextra = bpOptions[bpChoice].qextra;
+  ierr = PetscOptionsInt("-qextra", "Number of extra quadrature points",
+                         NULL, qextra, &qextra, NULL); CHKERRQ(ierr);
+  ierr = PetscOptionsString("-ceed", "CEED resource specifier",
+                            NULL, ceedresource, ceedresource,
+                            sizeof(ceedresource), NULL); CHKERRQ(ierr);
+  ierr = PetscOptionsString("-mesh", "Read mesh from file", NULL,
+                            filename, filename, sizeof(filename), &read_mesh);
+  CHKERRQ(ierr);
+  if (!read_mesh) {
+    PetscInt tmp = dim;
+    ierr = PetscOptionsIntArray("-cells","Number of cells per dimension", NULL,
+                                melem, &tmp, NULL); CHKERRQ(ierr);
+  }
+  enforce_bc = bpOptions[bpChoice].enforce_bc;
+  ierr = PetscOptionsBool("-enforce_bc", "Enforce essential BCs", NULL,
+                          enforce_bc, &enforce_bc, NULL); CHKERRQ(ierr);
+  ierr = PetscOptionsEnd(); CHKERRQ(ierr);
+
+  // Setup DM
+  if (read_mesh) {
+    ierr = DMPlexCreateFromFile(PETSC_COMM_WORLD, filename, PETSC_TRUE, &dm);
+    CHKERRQ(ierr);
+  } else {
+    ierr = DMPlexCreateBoxMesh(PETSC_COMM_WORLD, dim, PETSC_FALSE, melem, NULL,
+                                NULL, NULL, PETSC_TRUE ,&dm); CHKERRQ(ierr);
+  }
+
+  {
+    DM dmDist = NULL;
+    PetscPartitioner part;
+
+    ierr = DMPlexGetPartitioner(dm, &part);CHKERRQ(ierr);
+    ierr = PetscPartitionerSetFromOptions(part);CHKERRQ(ierr);
+    ierr = DMPlexDistribute(dm, 0, NULL, &dmDist);CHKERRQ(ierr);
+    if (dmDist) {
+      ierr = DMDestroy(&dm); CHKERRQ(ierr);
+      dm  = dmDist;
+    }
+  }
+  ierr = DMGetDimension(dm, &dim); CHKERRQ(ierr);
+  ierr = PetscFECreateDefault(PETSC_COMM_SELF, dim, vscale, PETSC_FALSE, NULL,
+                              PETSC_DETERMINE, &fe);
+  CHKERRQ(ierr);
+  ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
+  ierr = DMAddField(dm, NULL, (PetscObject)fe); CHKERRQ(ierr);
+  ierr = DMCreateDS(dm); CHKERRQ(ierr);
+  if (enforce_bc) {
+    marker_ids[0] = vscale;
+    ierr = DMAddBoundary(dm, DM_BC_ESSENTIAL, "wall", "marker", 0, 0, NULL,
```

Comment:
This is where the BP4/6 problem lies, I believe.

Reply:
Are the boundary conditions homogeneous or inhomogeneous? If inhomogeneous, you'll need `DMPlexInsertBoundaryValues`.
`DMGlobalToLocal` populates only the inessential dofs. That's usually what you want when applying a linear operator (leaving the others zero), but not when constructing a solution. Note that you may also need to lift those contributions to the RHS in this mode.

Reply:
We have homogeneous BCs. The BP 3/5 result is correct, so I think that there must be some issue when jumping from 1 component to 3.